tests/returntypes/test_autograd_qnode_new.py

```diff
@@ -0,0 +1,1942 @@
+
+        for args in spy.call_args_list:
+            assert args[1]["order"] == 2
+            assert args[1]["h"] == 1e-8
+
+    def test_changing_trainability(self, dev_name, diff_method, mode, mocker, tol):
+        """Test changing the trainability of parameters changes the
+        number of differentiation requests made"""
+        if diff_method != "parameter-shift":
+            pytest.skip("Test only supports parameter-shift")
+
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=True)
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qnode(dev, interface="autograd", diff_method="parameter-shift")
+        def circuit(a, b):
+            qml.RY(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(1))
+
+        def loss(a, b):
+            return np.sum(autograd.numpy.hstack(circuit(a, b)))
+
+        grad_fn = qml.grad(loss)
+        spy = mocker.spy(qml.gradients.param_shift, "transform_fn")
+        res = grad_fn(a, b)
+
+        # the tape has reported both arguments as trainable
+        assert circuit.qtape.trainable_params == [0, 1]
+
+        expected = [-np.sin(a) + np.sin(a) * np.sin(b), -np.cos(a) * np.cos(b)]
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # The parameter-shift rule has been called for each argument
+        assert len(spy.spy_return[0]) == 4
+
+        # make the second QNode argument a constant
+        a = np.array(0.54, requires_grad=True)
+        b = np.array(0.8, requires_grad=False)
+
+        res = grad_fn(a, b)
+
+        # the tape has reported only the first argument as trainable
+        assert circuit.qtape.trainable_params == [0]
+
+        expected = [-np.sin(a) + np.sin(a) * np.sin(b)]
+        print(res, expected)
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        # The parameter-shift rule has been called only once
+        assert len(spy.spy_return[0]) == 2
+
+        # trainability also updates on evaluation
+        a = np.array(0.54, requires_grad=False)
+        b = np.array(0.8, requires_grad=True)
+        circuit(a, b)
+        assert circuit.qtape.trainable_params == [1]
+
+    def test_classical_processing(self, dev_name, diff_method, mode, tol):
+        """Test classical processing within the quantum tape"""
+        a = np.array(0.1, requires_grad=True)
+        b = np.array(0.2, requires_grad=False)
+        c = np.array(0.3, requires_grad=True)
+
+        dev = qml.device(dev_name, wires=1)
+
+        @qnode(dev, diff_method=diff_method, interface="autograd", mode=mode)
+        def circuit(a, b, c):
+            qml.RY(a * c, wires=0)
+            qml.RZ(b, wires=0)
+            qml.RX(c + c**2 + np.sin(a), wires=0)
+            return qml.expval(qml.PauliZ(0))
+
+        res = qml.jacobian(circuit)(a, b, c)
+
+        if diff_method == "finite-diff":
```

Comment:
Is this not the case for param-shift?

Reply:
I have no idea why it was here in the first place, let me double check.

Reply:
Removed!